src/fairseq2/nn/lora.py

```diff
@@ -0,0 +1,373 @@
+            continue
+
+        submodule_path = name.split(".")
+        parent = module.get_submodule(".".join(submodule_path[:-1]))
+        submodule_name = submodule_path[-1]
+
+        unwrapped_layer: Optional[nn.Module] = None
+        if isinstance(submodule, LoRALinear):
+            # TODO: currently there's no way to distinguish which type the
+            # original module is (`Linear` or `TiedProjection` or
+            # `QKVProjection`). Using `Linear` should be functionally
+            # identical (output would be the same), but since it might be a
+            # different projection layer, it might cause issues in
+            # downstream operations.
+            unwrapped_layer = Linear(
+                submodule.input_dim,
+                submodule.output_dim,
+                bias=submodule.bias is not None,
+                skip_init=True,
+                device=submodule.weight.device,  # type: ignore[arg-type]
+                dtype=submodule.weight.dtype,  # type: ignore[arg-type]
+            )
+            unwrapped_layer.weight = submodule.weight
+            if submodule.bias is not None:
+                unwrapped_layer.bias = submodule.bias
+        elif isinstance(submodule, LoRAEmbedding):
+            unwrapped_layer = StandardEmbedding(
+                num_embeddings=submodule.num_embeddings,
+                embedding_dim=submodule.embedding_dim,
+                pad_idx=submodule.pad_idx,
+                device=submodule.weight.device,  # type: ignore[arg-type]
+                dtype=submodule.weight.dtype,  # type: ignore[arg-type]
+            )
+            unwrapped_layer.weight = submodule.weight
+        else:
+            raise ValueError(
+                f"Cannot unwrap the module '{name}' as the module type '{type(submodule).__name__}' is not supported."
+            )
+
+        unwrapped_layer.train(mode=submodule.training)
+        setattr(parent, submodule_name, unwrapped_layer)
+
+    return module
+
+
+def merge_lora(module: nn.Module) -> None:
+    # Iterate through all `LoRALayer`s in `module`, and call their `merge`
+    # method which is expected to merge LoRA A, B weights with the wrapped W.
+    for submodule in module.modules():
+        if isinstance(submodule, LoRALayer):
+            submodule.merge()
+
+
+def unmerge_lora(module: nn.Module) -> None:
+    for submodule in module.modules():
+        if isinstance(submodule, LoRALayer):
+            submodule.unmerge()
+
+
+def lora_state_dict(module: nn.Module) -> Dict[str, Any]:
+    state_dict = module.state_dict()
+    lora_states = {name: state for name, state in state_dict.items() if "lora_" in name}
+    return lora_states
+
+
+def freeze_non_lora(
+    module: nn.Module, unfreeze_bias: Literal["none", "all", "lora_only"] = "none"
+) -> None:
+    # Set requires_grad to False for all parameters in the module except
+    # lora layers
+    for name, param in module.named_parameters():
```

Comment:
Similarly, this is also error prone. I would do it in semi-pseudo code:

```python
for name, submodule in module.named_modules():
   if not isinstance(submodule, LoRALayer):
      for param_name, param in submodule.named_parameters(recurse=False):
         if param_name == "bias":
            ...
         else:
           param.requires_grad = False
```

Reply:
Updated